### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by AmrMekhemar**

### CODE REVIEW

File: week7/NYTArticles/app/src/main/java/com/example/nytarticles/App.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        private val articleDb: ArticleDB by lazy {
+            Room.databaseBuilder(appInstance, ArticleDB::class.java, DATABASE_NAME)
+                .build()
+        }
+        private val articlesDAO: ArticlesDAO by lazy {
+            articleDb.articleDao()
+        }
+        val BASE_URL = "https://api.nytimes.com/svc/topstories/v2/"
+        private val apiService by lazy { buildApiService() }
+        val articleRepo by lazy { ArticleRepo(apiService, articlesDAO) }
```
Comment (filbabic) [inline week7/NYTArticles/app/src/main/java/com/example/nytarticles/App.kt:24]: Good job regarding the setup here!

So many people don't seem to understand they can do this, to reuse their dependencies! :]

File: week7/NYTArticles/app/src/main/java/com/example/nytarticles/model/Article.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    var byline: String = "",
+    var item_type: String = "",
+    var updated_date: String = "",
+    var created_date: String = "",
+    var published_date: String = "",
+    var material_type_facet: String = "",
+    var multimedia: List<Multimedia> = listOf(),
+    var short_url: String = "",
+    var isFavorite: Boolean = false
+
```
Comment (filbabic) [inline week7/NYTArticles/app/src/main/java/com/example/nytarticles/model/Article.kt:26]: Make sure to remove extra space wherever possible!

File: week7/NYTArticles/app/src/main/java/com/example/nytarticles/repository/ArticleRepo.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import com.example.nytarticles.BuildConfig.api_key
+import com.example.nytarticles.database.ArticlesDAO
+import com.example.nytarticles.model.Article
+import com.example.nytarticles.networking.RemoteApiService
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.withContext
+
+
+class ArticleRepo(private val apiService: RemoteApiService, private val articlesDAO: ArticlesDAO) {
```
Comment (filbabic) [inline week7/NYTArticles/app/src/main/java/com/example/nytarticles/repository/ArticleRepo.kt:13]: Really nice passing in the deps through the constructor!

File: week7/NYTArticles/app/src/main/java/com/example/nytarticles/utils/ArticleAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import com.bumptech.glide.Glide
+import com.example.nytarticles.App
+import com.example.nytarticles.R
+import com.example.nytarticles.model.Article
+
+
+class ArticleAdapter(private var articles: List<Article>, val listener: (Any) -> Unit) :
```
Comment (filbabic) [inline week7/NYTArticles/app/src/main/java/com/example/nytarticles/utils/ArticleAdapter.kt:15]: Can you change the Any type to be more concrete?

File: week7/NYTArticles/app/src/main/java/com/example/nytarticles/view/ArticlesFragment.kt [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                    it.isFavorite = true
+                    articlesViewModel.saveFavoriteArticleToDB(it)
+                    requireContext().toast(getString(R.string.added_to_favorites))
+                } else {
+                    it.isFavorite = false
+                    articlesViewModel.saveFavoriteArticleToDB(it)
+                    requireContext().toast(getString(R.string.removed_from_favorites))
+                }
+
+            } else {
```
Comment (filbabic) [inline week7/NYTArticles/app/src/main/java/com/example/nytarticles/view/ArticlesFragment.kt:121]: Okay this is why you have Any.

maybe have two different listeners then, to have the code a bit more separate?

**Review by filbabic — APPROVED** [review]
Really nice and well structured solution! I love how you have all these advanced features like Flow, and how you've already applied SOLID to great extent.

**Above & Beyond**. I'd love for you to showcase your project in the next session, for Week 8, just to show your work! :]

**Comment by AmrMekhemar** [comment]
Yeah,  okay...
Thanks so much!

**Pull request merged**